### ISSUE TITLE
Implement PartialEq for Map or MapFactory

### ISSUE THREAD

**Issue by thebearjew** [yukinarit/mapbox-gl-rs#11]
When using `mapboxgl` inside of a yew app, you might want to pass around `MapFactory` as props to other components so they can manipulate the map in different ways.

Additionally, in order to take a map instance and pass it into a Yew context provider like `use_context`, the type must implement `ParitalEq` (docs for [use_context](https://docs.rs/yew/0.20.0/yew/functional/fn.use_context.html)).

Global state management crates for Yew like [bounce](https://bounce-rs.org) also require that the shared value be `ParitalEq`.

**Comment by yukinarit** [comment]
Hi @thebearjew 

I am not an expert on this area so please correct me if this doesn't make any sense.

I think `PartialEq` is required to propagate changes on properties between components.  `Map` has its own event listener, so you can hook into events that occurred inside Map. If you want to just pass around `Map` in components and call Map's methods or properties, It doesn't make sense to implement "deep object comparison" (e.g. [How to Compare Objects in JavaScript](https://dmitripavlutin.com/how-to-compare-objects-in-javascript/))? 🤔  

That said, implementing `PartialEq` to compare underlying JavaScript Object references is enough? 🤔 

`Handle`
```rust
impl PartialEq for Handle {
    fn eq(&self, other: &Self) -> bool {
        js_sys::Object::is(self.on_data.as_ref(), other.on_data.as_ref()) &&
        js_sys::Object::is(self.on_styledata.as_ref(), other.on_styledata.as_ref()) &&
        ...
    }
}
```

`Map` and `MapFactory`
```rust
#[derive(PartialEq)]
pub struct MapFactory {
    pub map: Arc<Map>,
    handle: Option<Handle>,
}

pub struct Map {
    pub(crate) inner: crate::js::Map,
}

impl PartialEq for Map {
    fn eq(&self, other: &Self) -> bool {
        js_sys::Object::is(&self.inner, &other.inner)
    }
}
```

**NOTE**: Above code compiles but wasn't tested at all.

**Comment by thebearjew** [comment]
I agree that using JavaScript referential equality (`Object::is`) is enough, and I have a PR I am working on for this which also introduces `Clone`.

For some context, the scenario I am trying to implement is one in which I can control the behavior of the map via Yew hooks and lifecycle events, rather than respond to Mapbox events like `onload` and `onclick`. Additionally, when user interactions like button clicks, route changes, and external data are loaded, I would like to control the map to transition it. All of which require that `Map` and `MapFactory` be have traits which Yew require like [ContextProvider](https://docs.rs/yew/latest/yew/context/struct.ContextProvider.html#).

Example:

```rust

#[function_component(LoadRemoteGeoJson)]
fn load_remote_geojson(props: Props) -> Html {
    let geojson = use_fetch::<MyGeoJsonData>(props.url);
    let map = use_context::<Map>();

   map.add_geojeon_source("walking", geojson);
   map.add_layer("walking", ...);

   // renders props.children
   html! { ... }
}

fn switch(route: Route) -> Html {
    match route {
        Route::User { id } => html {
            // 
            <LoadRemoteGeoJson url={format!("https://example.com/{id}.geojson")}>
                <UserProfile {id} />
            </LoadRemoteGeoJson>
        },
    }
}

#[function_component(App)]
fn app() -> Html {
    // Initialize the map. copied mapbox-gl-rs example.
    let map  = use_map();
    html! {
        // Make the `MapFactory` or `Map` available to any component in the application.
        // This requires `Clone` and `PartialEq` on `Map` or `MapFactory`
        <ContextProvider<Map> context={map.clone}>
            <BrowserRouter>
                // Load different GeoJson based on route
                <Switch<Route> render={switch} />
            </BrowserRouter>
        </ContextProvider<Map>>
    }
 
}
```
